experimental/hip/hip_buffer.c

```diff
@@ -0,0 +1,167 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "experimental/hip/hip_buffer.h"
+
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "iree/base/api.h"
+#include "iree/base/tracing.h"
+
+typedef struct iree_hal_hip_buffer_t {
+  hipDeviceptr_t device_ptr;
+  iree_hal_buffer_t base;
+  iree_hal_buffer_release_callback_t release_callback;
+  iree_hal_hip_buffer_type_t type;
+  void* host_ptr;
+} iree_hal_hip_buffer_t;
+
+static const iree_hal_buffer_vtable_t iree_hal_hip_buffer_vtable;
+
+static iree_hal_hip_buffer_t* iree_hal_hip_buffer_cast(
+    iree_hal_buffer_t* base_value) {
+  IREE_HAL_ASSERT_TYPE(base_value, &iree_hal_hip_buffer_vtable);
+  return (iree_hal_hip_buffer_t*)base_value;
+}
+
+static const iree_hal_hip_buffer_t* iree_hal_hip_buffer_const_cast(
+    const iree_hal_buffer_t* base_value) {
+  IREE_HAL_ASSERT_TYPE(base_value, &iree_hal_hip_buffer_vtable);
+  return (const iree_hal_hip_buffer_t*)base_value;
+}
+
+iree_status_t iree_hal_hip_buffer_wrap(
+    iree_hal_allocator_t* allocator, iree_hal_memory_type_t memory_type,
+    iree_hal_memory_access_t allowed_access,
+    iree_hal_buffer_usage_t allowed_usage, iree_device_size_t allocation_size,
+    iree_device_size_t byte_offset, iree_device_size_t byte_length,
+    iree_hal_hip_buffer_type_t buffer_type, hipDeviceptr_t device_ptr,
+    void* host_ptr, iree_hal_buffer_release_callback_t release_callback,
+    iree_allocator_t host_allocator, iree_hal_buffer_t** out_buffer) {
+  IREE_ASSERT_ARGUMENT(out_buffer);
+  if (!host_ptr && iree_any_bit_set(allowed_usage,
+                                    IREE_HAL_BUFFER_USAGE_MAPPING_PERSISTENT |
+                                        IREE_HAL_BUFFER_USAGE_MAPPING_SCOPED)) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "mappable buffers require host pointers");
+  }
+
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_hip_buffer_t* buffer = NULL;
+  iree_status_t status =
+      iree_allocator_malloc(host_allocator, sizeof(*buffer), (void**)&buffer);
+  if (iree_status_is_ok(status)) {
+    iree_hal_buffer_initialize(host_allocator, allocator, &buffer->base,
+                               allocation_size, byte_offset, byte_length,
+                               memory_type, allowed_access, allowed_usage,
+                               &iree_hal_hip_buffer_vtable, &buffer->base);
+    buffer->type = buffer_type;
```

Comment:
We'd want this order to follow the definition order so it's easier to check which one is missing or add new ones. I'd recommnad to fix the definition order to follow the order here.